Legg til typehints

Python krever ikke typehints, men det er en effektiv måte å skrive selvdokumenterende kode på. 
Bruk gjerne `from typing import ...` for komplekse typer som `List`, `Dict`, `Union`, `Callable`, `Iterable`, `Any` osv.

`networkx` har ikke type hints, så mypy varsler om manglende stub file: 

```
dhlab/api/nb_ngram_api.py:4: error: Skipping analyzing "networkx": module is installed, but missing library stubs or py.typed marker  [import]
dhlab/api/nb_ngram_api.py:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

Jeg legger til `# type: ignore` på linjene som importerer `networkx` for å ignorere mypy-varselet.

mypy-loggen etter commit https://github.com/NationalLibraryOfNorway/DHLAB/commit/a6823b43906d16fc1e3ccfbb83dc5656234d8d62 (på [docs/setup-markdown-docs](https://github.com/NationalLibraryOfNorway/DHLAB/tree/docs/setup-markdown-docs))


```
$ mypy dhlab/
dhlab/nbtokenizer.py:417: error: Incompatible types in assignment (expression has type "Pattern[str]", variable has type "list[str]")  [assignment]
dhlab/api/nb_ngram_api.py:53: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
dhlab/api/nb_ngram_api.py:54: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
dhlab/nbpictures.py:158: error: Name "Optional" is not defined  [name-defined]
dhlab/nbpictures.py:158: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Optional")
dhlab/ngram/nb_ngram.py:28: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
dhlab/ngram/nb_ngram.py:57: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
dhlab/ngram/nb_ngram.py:57: note: Possible overload variants:
dhlab/ngram/nb_ngram.py:57: note:     def __getitem__(self, SupportsIndex, /) -> Any
dhlab/ngram/nb_ngram.py:57: note:     def __getitem__(self, slice, /) -> list[Any]
dhlab/ngram/nb_ngram.py:59: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
dhlab/ngram/nb_ngram.py:59: note: Possible overload variants:
dhlab/ngram/nb_ngram.py:59: note:     def __getitem__(self, SupportsIndex, /) -> Any
dhlab/ngram/nb_ngram.py:59: note:     def __getitem__(self, slice, /) -> list[Any]
dhlab/api/nb_search_api.py:46: error: Argument "params" to "get" has incompatible type "dict[str, object]"; expected "SupportsItems[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | Iterable[tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None]] | str | bytes | None"  [arg-type]
dhlab/api/nb_search_api.py:51: error: Argument "params" to "get" has incompatible type "dict[str, object]"; expected "SupportsItems[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | Iterable[tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None]] | str | bytes | None"  [arg-type]
dhlab/api/nb_search_api.py:76: error: Argument "params" to "get" has incompatible type "dict[str, object]"; expected "SupportsItems[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | Iterable[tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None]] | str | bytes | None"  [arg-type]
dhlab/api/dhlab_api.py:20: error: Argument "params" to "get" has incompatible type "dict[str, object]"; expected "SupportsItems[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None] | Iterable[tuple[str | bytes | int | float, str | bytes | int | float | Iterable[str | bytes | int | float] | None]] | str | bytes | None"  [arg-type]
dhlab/api/dhlab_api.py:234: error: Item "None" of "list[Any] | None" has no attribute "__iter__" (not iterable)  [union-attr]
dhlab/api/dhlab_api.py:466: error: Unsupported operand types for + ("str" and "None")  [operator]
dhlab/api/dhlab_api.py:466: note: Right operand is of type "str | None"
dhlab/api/dhlab_api.py:470: error: "Index" has no attribute "levels"; maybe "nlevels"?  [attr-defined]
dhlab/api/dhlab_api.py:520: error: "Index" has no attribute "levels"; maybe "nlevels"?  [attr-defined]
dhlab/api/dhlab_api.py:568: error: "Index" has no attribute "levels"; maybe "nlevels"?  [attr-defined]
dhlab/api/dhlab_api.py:605: error: "Index" has no attribute "levels"; maybe "nlevels"?  [attr-defined]
dhlab/api/dhlab_api.py:640: error: Argument "by" to "sort_values" of "DataFrame" has incompatible type "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex"; expected "str | Sequence[str]"  [arg-type]
dhlab/api/dhlab_api.py:643: error: Incompatible types in assignment (expression has type "list[str]", variable has type "Index")  [assignment]
dhlab/api/dhlab_api.py:687: error: Incompatible types in assignment (expression has type "list[str]", variable has type "Index")  [assignment]
dhlab/api/dhlab_api.py:835: error: Incompatible return value type (got "dict[<nothing>, <nothing>]", expected "DataFrame")  [return-value]
dhlab/api/dhlab_api.py:866: error: Incompatible return value type (got "dict[<nothing>, <nothing>]", expected "DataFrame")  [return-value]
dhlab/images/nbpictures.py:83: error: Argument 1 to "iiif_manifest" has incompatible type "int | str"; expected "str"  [arg-type]
dhlab/text/concordance.py:32: error: Incompatible types in assignment (expression has type "DataFrame", variable has type "Styler")  [assignment]
dhlab/text/concordance.py:51: error: Incompatible types in assignment (expression has type "Series[Any]", variable has type "Index")  [assignment]
dhlab/text/concordance.py:54: error: Incompatible return value type (got "HTML", expected "DataFrame")  [return-value]
dhlab/text/collocations.py:59: error: Argument "by" to "sort_values" of "DataFrame" has incompatible type "str | None"; expected "str | Sequence[str]"  [arg-type]
dhlab/text/collocations.py:59: error: Argument 1 to "head" of "DataFrame" has incompatible type "int | None"; expected "int"  [arg-type]
Found 28 errors in 9 files (checked 28 source files)
```
Mange av disse feilene vil kreve en del refaktorering å rette opp i. 
Vi bør lage regresjonstester for funksjonene det gjelder før vi endrer på koden.

--- a/.gitignore
+++ b/.gitignore
@@ -164,3 +164,13 @@ dhlab/_version.py
 # documentation legacy files
 docs/new_conf.py
 docs/orig_conf.py
+
+# Devenv
+.devenv*
+devenv.local.nix
+
+# direnv
+.direnv
+
+# pre-commit
+.pre-commit-config.yaml

--- a/dhlab/text/chunking.py
+++ b/dhlab/text/chunking.py
@@ -6,7 +6,7 @@ from dhlab.api.dhlab_api import get_chunks, get_chunks_para
 class Chunks:
     """Create chunks from a text."""
 
-    def __init__(self, urn=None, chunks=1000):
+    def __init__(self, urn: str | None = None, chunks: int = 1000):
         """
         :param urn: str or list
         :param chunks: {'para', 'avsn'} or int

--- a/dhlab/text/conc_coll.py
+++ b/dhlab/text/conc_coll.py
@@ -53,13 +53,12 @@ class Concordance(DhlabObj):
         super().__init__(self.concordance)
 
     def show(self, n=10, style=True):
+        result = self.concordance.sample(min(n, self.size))
+
         if style:
-            result = self.concordance.sample(min(n, self.size))[
-                ["link", "concordance"]
-            ].style
+            return result[["link", "concordance"]].style
         else:
-            result = self.concordance.sample(min(n, self.size))
-        return result
+            return result
 
     @classmethod
     def from_df(cls, df):

--- a/dhlab/text/conc_coll.py
+++ b/dhlab/text/conc_coll.py
@@ -10,6 +10,7 @@ from dhlab.api.dhlab_api import (
     concordance_counts,
 )
 from dhlab.text.dhlab_object import DhlabObj
+from dhlab.text.corpus import Corpus
 from dhlab.text.utils import urnlist
 
 
@@ -27,7 +28,13 @@ def find_hits(x):
 class Concordance(DhlabObj):
     """Wrapper for concordance function"""
 
-    def __init__(self, corpus=None, query=None, window=20, limit=500):
+    def __init__(
+        self,
+        corpus: Corpus | None = None,
+        query: str | None = None,
+        window: int = 20,
+        limit: int = 500
+):
         """Get concordances for word(s) in corpus
 
         :param corpus: Target corpus, defaults to None
@@ -48,11 +55,10 @@ class Concordance(DhlabObj):
             self.concordance = self.concordance[["link", "urn", "conc"]]
             self.concordance.columns = ["link", "urn", "concordance"]
             self.corpus = corpus
-        # self.size = len(self.concordance)
 
         super().__init__(self.concordance)
 
-    def show(self, n=10, style=True):
+    def show(self, n: int = 10, style: bool = True):
         result = self.concordance.sample(min(n, self.size))
 
         if style:
@@ -61,7 +67,7 @@ class Concordance(DhlabObj):
             return result
 
     @classmethod
-    def from_df(cls, df):
+    def from_df(cls, df: pd.DataFrame):
         "Typecast DataFrame to Concordance"
         obj = Concordance()
         obj.concordance = df
@@ -74,14 +80,14 @@ class Collocations(DhlabObj):
 
     def __init__(
         self,
-        corpus=None,
-        words=None,
-        before=10,
-        after=10,
-        reference=None,
-        samplesize=20000,
-        alpha=False,
-        ignore_caps=False,
+        corpus: Corpus | None = None,
+        words: str | list[str] | None = None,
+        before: int = 10,
+        after: int = 10,
+        reference: pd.DataFrame | None = None,
+        samplesize: int = 20000,
+        alpha: bool = False,
+        ignore_caps: bool = False,
     ):
         """Create collocations object
 
@@ -136,23 +142,23 @@ class Collocations(DhlabObj):
         self.before = before
         self.after = after
 
-        if reference is not None:
+        if self.reference is not None:
             teller = self.coll.counts / self.coll.counts.sum()
             divisor = self.reference.iloc[:, 0] / self.reference.iloc[:, 0].sum()
             self.coll["relevance"] = teller / divisor
 
         super().__init__(self.coll)
 
-    def show(self, sortby="counts", n=20):
+    def show(self, sortby: str = "counts", n: int = 20):
         return self.coll.sort_values(by=sortby, ascending=False).head(n)
 
-    def keywordlist(self, top=200, counts=5, relevance=10):
+    def keywordlist(self, top: int = 200, counts: int = 5, relevance: float = 10):
         mask = self.coll[self.coll.counts > counts]
         mask = mask[mask.relevance > relevance]
         return list(mask.sort_values(by="counts", ascending=False).head(200).index)
 
     @classmethod
-    def from_df(cls, df):
+    def from_df(cls, df: pd.DataFrame):
         """Typecast DataFrame to Collocation
 
         :param df: DataFrame
@@ -167,7 +173,13 @@ class Collocations(DhlabObj):
 class Counts(DhlabObj):
     """Provide counts for a corpus - shouldn't be too large"""
 
-    def __init__(self, corpus=None, words=None, cutoff=0, sparse=True):
+    def __init__(
+        self,
+        corpus: Corpus | None = None,
+        words: list[str] | None = None,
+        cutoff: int = 0,
+        sparse: bool = True
+):
         """Get frequency list for Corpus
 
         :param corpus: target Corpus, defaults to None
@@ -259,7 +271,7 @@ class Counts(DhlabObj):
         return self.relfreq.rename(self.title_dct, axis=1)
 
     @classmethod
-    def from_df(cls, df):
+    def from_df(cls, df: pd.DataFrame):
         obj = Counts()
         obj.freq = df
         obj.frame = df

--- a/dhlab/text/corpus.py
+++ b/dhlab/text/corpus.py
@@ -32,26 +32,26 @@ class Corpus(DhlabObj):
 
     def __init__(
         self,
-        doctype=None,
-        author=None,
-        freetext=None,
-        fulltext=None,
-        from_year=None,
-        to_year=None,
-        from_timestamp=None,
-        to_timestamp=None,
-        title=None,
-        ddk=None,
-        subject=None,
-        publisher=None,
-        literaryform=None,
-        genres=None,
-        city=None,
-        lang=None,
-        limit=10,
-        limit_by_year=False,
-        order_by="random",
-        allow_duplicates=False,
+        doctype: str | None = None,
+        author: str | None = None,
+        freetext: str | None = None,
+        fulltext: str | None = None,
+        from_year: int | None = None,
+        to_year: int | None = None,
+        from_timestamp: int | None = None,
+        to_timestamp: int | None = None,
+        title: str | None = None,
+        ddk: str | None = None,
+        subject: str | None = None,
+        publisher: str | None = None,
+        literaryform: str | None = None,
+        genres: str | None = None,
+        city: str | None = None,
+        lang: str | None = None,
+        limit: int | None = 10,
+        limit_by_year: bool = False,
+        order_by: str | None = "random",
+        allow_duplicates: bool = False,
     ):
         """Create Corpus
 
@@ -221,7 +221,7 @@ class Corpus(DhlabObj):
         new_corpus = get_metadata(urnlist(identifiers))
         self.add(new_corpus)
 
-    def evaluate_words(self, wordbags=None):
+    def evaluate_words(self, wordbags: dict | None = None):
         df = evaluate_documents(wordbags=wordbags, urns=list(self.frame.urn))
         df.index = df.index.astype(int)
         cols = df.columns
@@ -254,7 +254,7 @@ class Corpus(DhlabObj):
         mask = self.frame.langs.apply(lambda x: len(x.split("/"))) == 1
         return self.from_df(self.frame[mask])
 
-    def conc(self, words, window: int = 20, limit: int = 500) -> dh.Concordance:
+    def conc(self, words: str | None, window: int = 20, limit: int = 500) -> dh.Concordance:
         """Get concodances of `words` in corpus"""
         return dh.Concordance(
             corpus=self.frame, query=words, window=window, limit=limit
@@ -262,13 +262,13 @@ class Corpus(DhlabObj):
 
     def coll(
         self,
-        words=None,
-        before=10,
-        after=10,
-        reference=None,
-        samplesize=20000,
-        alpha=False,
-        ignore_caps=False,
+        words: str | list[str] | None = None,
+        before: int = 10,
+        after: int = 10,
+        reference: pd.DataFrame | None = None,
+        samplesize: int = 20000,
+        alpha: bool = False,
+        ignore_caps: bool = False,
     ) -> dh.Collocations:
         """Get collocations of `words` in corpus"""
         return dh.Collocations(
@@ -282,14 +282,15 @@ class Corpus(DhlabObj):
             ignore_caps=ignore_caps,
         )
 
-    def count(self, words=None, cutoff=0, sparse=True):
+    def count(self, words: list[str] | None = None, cutoff: int = 0, sparse: bool = True):
         """Get word frequencies for corpus"""
         return dh.Counts(self, words, cutoff, sparse)
 
-    def freq(self, words=None, cutoff=0, sparse=True):
+    def freq(self, words: list[str] | None = None, cutoff: int = 0, sparse: bool = True):
         """Get word frequencies for corpus"""
         return dh.Counts(self, words, cutoff, sparse)
 
+    # TODO: Is this `corpus: "Corpus"` annotation style deliberate? We're already importing __future__.annotations.
     @staticmethod
     def _is_Corpus(corpus: "Corpus") -> bool:
         """Check if `input` is Corpus or DataFrame"""
@@ -297,7 +298,8 @@ class Corpus(DhlabObj):
             raise TypeError("Input is not Corpus or DataFrame")
         return isinstance(corpus, Corpus) | isinstance(corpus, DataFrame)
 
-    def __add__(self, other):
+    # TODO: Is this `corpus: "Corpus"` annotation style deliberate? We're already importing __future__.annotations.
+    def __add__(self, other: "Corpus"):
         """Add two Corpus objects"""
         if not self._is_Corpus(other):
             raise TypeError("Input is not Corpus or DataFrame")

--- a/dhlab/text/parse.py
+++ b/dhlab/text/parse.py
@@ -11,7 +11,13 @@ class Models:
 class NER:
     """Provide NER"""
 
-    def __init__(self, urn=None, model=None, start_page=0, to_page=0):
+    def __init__(
+        self,
+        urn: str | None = None,
+        model: str | None = None,
+        start_page: int = 0,
+        to_page: int = 0
+    ):
         self.model = model
         self.ner = ner_from_urn(
             urn=urn, model=self.model, start_page=start_page, to_page=to_page
@@ -21,7 +27,13 @@ class NER:
 class POS:
     """Provide POS and a parse"""
 
-    def __init__(self, urn=None, model=None, start_page=0, to_page=0):
+    def __init__(
+        self,
+        urn: str | None = None,
+        model: str | None = None,
+        start_page: int = 0,
+        to_page: int = 0
+    ):
         self.model = model
         self.pos = pos_from_urn(
             urn=urn, model=self.model, start_page=start_page, to_page=to_page

